federatedscope/attack/trainer/benign_trainer.py

```diff
@@ -0,0 +1,223 @@
+#             ctx.cfg)
+#         setattr(ctx, "poison_{}_loader".format(ctx.cur_data_split),
+#                 ReIterator(loader))
+#     elif not isinstance(ctx.get("poison_{}_loader".format(ctx.cur_data_split)),
+#                         ReIterator):
+#         setattr(
+#             ctx, "poison_{}_loader".format(ctx.cur_data_split),
+#             ReIterator(ctx.get("poison_{}_loader".format(ctx.cur_data_split))))
+#     else:
+#         ctx.get("poison_{}_loader".format(ctx.cur_data_split)).reset()
+        
+
+
+# def hook_on_batch_start_test_poison(ctx):
+#     # prepare data batch
+#     try:
+#         ctx.poison_data_batch = next(
+#             ctx.get("poison_{}_loader".format(ctx.cur_data_split)))
+#     except StopIteration:
+#         raise StopIteration
+
+
+# def hook_on_batch_forward_test_poison(ctx):
+
+#     x, label = [_.to(ctx.device) for _ in ctx.poison_data_batch]
+#     pred = ctx.model(x)
+#     if len(label.size()) == 0:
+#         label = label.unsqueeze(0)
+#     ctx.poison_loss_batch = ctx.criterion(pred, label)
+#     ctx.poison_y_true = label
+#     ctx.poison_y_prob = pred
+
+#     ctx.poison_batch_size = len(label)
+
+
+
+# def hook_on_batch_end_test_poison(ctx):
+
+#     # update statistics
+#     setattr(
+#         ctx, "poison_loss_batch_total_{}".format(ctx.cur_data_split),
+#         ctx.get("poison_loss_batch_total_{}".format(ctx.cur_data_split)) +
+#         ctx.poison_loss_batch.item() * ctx.poison_batch_size)
+
+
+#     setattr(
+#         ctx, "poison_num_samples_{}".format(ctx.cur_data_split),
+#         ctx.get("poison_num_samples_{}".format(ctx.cur_data_split)) +
+#         ctx.poison_batch_size)
+
+#     # cache label for evaluate
+#     ctx.get("poison_{}_y_true".format(ctx.cur_data_split)).append(
+#         ctx.poison_y_true.detach().cpu().numpy())
+
+#     ctx.get("poison_{}_y_prob".format(ctx.cur_data_split)).append(
+#         ctx.poison_y_prob.detach().cpu().numpy())
+
+#     # clean temp ctx
+#     ctx.poison_data_batch = None
+#     ctx.poison_batch_size = None
+#     ctx.poison_loss_task = None
+#     ctx.poison_loss_batch = None
+#     ctx.poison_loss_regular = None
+#     ctx.poison_y_true = None
+#     ctx.poison_y_prob = None
+
+
+
+def hook_on_fit_end_test_poison(ctx):
+
+    """Evaluate metrics of poisoning attacks.
+
+    """
+
+    ctx['poison_'+ctx.cur_data_split+'_loader'] = ctx.data['poison_'+ctx.cur_data_split]
+    ctx['poison_'+ctx.cur_data_split+'_data'] = ctx.data['poison_'+ctx.cur_data_split].dataset
+    ctx['num_poison_'+ctx.cur_data_split+'_data'] = len(ctx.data['poison_'+ctx.cur_data_split].dataset)
+    setattr(ctx, "poison_{}_y_true".format(ctx.cur_data_split), [])
+    setattr(ctx, "poison_{}_y_prob".format(ctx.cur_data_split), [])
+    setattr(ctx, "poison_num_samples_{}".format(ctx.cur_data_split), 0)
+
+    for batch_idx, (samples, targets) in enumerate(ctx['poison_'+ctx.cur_data_split+'_loader']):
```

Comment:
To conduct evaluation on the "poison" dataset for each client after a training routine, you can use the existing APIs as follows:
- Add the "poison" data into the global data dict, like
```
{
    '1': {
        "train": ...
        "test": ...
        "val": ...
        "posion": ...
    },
    '2': {
        "train": ...
        "test": ...
        "val": ...
        "posion": ...
    },
    ...
}
```
- Set `config.eval.split=["posion"]` to specifies the part of dataset for evaluation. If your choice metric is accuracy (`config.eval.metric=["acc"]`), then the evaluation results will be reported by the name of seleted split "posion" and metric "acc": "posion_acc"
- Set `config.eval.freq` as what you want.

By this way,  you don't need to code a test routine within the trainer and worry about duplicate variables within the context `ctx`.

Reply:
fixed it